crates/rattler_repodata_gateway/src/gateway/query.rs

```diff
@@ -117,46 +144,75 @@ impl GatewayQuery {
+                    if let Some(record) = record.first() {
+                        let spec_name = spec
+                            .0
+                            .clone()
+                            .name
+                            .ok_or(GatewayError::MatchSpecWithoutName(spec.0.clone()))?;
+                        if record.package_record.name != spec_name {
+                            // Using as_source to get the closest to the retrieved input.
+                            return Err(GatewayError::NotMatchingNameUrl(
+                                record.package_record.name.as_source().to_string(),
+                                spec_name.as_source().to_string(),
+                            ));
+                        }
+                    }
+                    // Push the direct url in the first subdir result for channel priority logic.
+                    Ok((0, vec![spec.0], record))
+                }
+                .boxed(),
+            );
+        }
+
+        let len = subdirs.len() + direct_url_offset;
+        let mut result = vec![RepoData::default(); len];
 
         // Loop until all pending package names have been fetched.
         loop {
-            // Iterate over all pending package names and create futures to fetch them from all
-            // subdirs.
+            // Iterate over all pending package names and create futures to fetch them from
+            // all subdirs.
             for (package_name, specs) in pending_package_specs.drain() {
                 for (subdir_idx, subdir) in subdirs.iter().cloned() {
                     let specs = specs.clone();
                     let package_name = package_name.clone();
                     let reporter = self.reporter.clone();
-                    pending_records.push(async move {
-                        let barrier_cell = subdir.clone();
-                        let subdir = barrier_cell.wait().await;
-                        match subdir.as_ref() {
-                            Subdir::Found(subdir) => subdir
-                                .get_or_fetch_package_records(&package_name, reporter)
-                                .await
-                                .map(|records| (subdir_idx, specs, records)),
-                            Subdir::NotFound => Ok((subdir_idx, specs, Arc::from(vec![]))),
+                    pending_records.push(
+                        async move {
+                            let barrier_cell = subdir.clone();
+                            let subdir = barrier_cell.wait().await;
+                            match subdir.as_ref() {
+                                Subdir::Found(subdir) => subdir
+                                    .get_or_fetch_package_records(&package_name, reporter)
+                                    .await
+                                    .map(|records| (subdir_idx, specs, records)),
+                                Subdir::NotFound => Ok((subdir_idx, specs, Arc::from(vec![]))),
```

Comment:
```suggestion
                                Subdir::NotFound => Ok((subdir_idx+1, specs, Arc::from(vec![]))),
```

Reply:
fixed in : https://github.com/mamba-org/rattler/pull/725/commits/44c28640e5e2193fe0321bee15eccc13735c734